deluge/httpdownloader.py

```diff
@@ -52,43 +79,55 @@ def __init__(
         :param headers: any optional headers to send
         :type headers: dictionary
         """
+        if not headers:
+            headers = {}
+        else:
+            for key, value in headers.items():
+                headers[key] = [value]
+        headers.update({'User-Agent': ['Deluge/%s (http://deluge-torrent.org)' % get_version()]})
 
+        self.url = url.encode()
+        self.filename = filename
         self.part_callback = part_callback
-        self.current_length = 0
-        self.total_length = 0
-        self.decoder = None
-        self.value = filename
+        self.headers = Headers(headers)
         self.force_filename = force_filename
         self.allow_compression = allow_compression
-        self.code = None
-        agent = 'Deluge/%s (http://deluge-torrent.org)' % get_version()
-        client.HTTPDownloader.__init__(
-            self, url, filename, headers=headers, agent=agent.encode('utf-8'))
-
-    def gotHeaders(self, headers):  # NOQA: N802
-        self.code = int(self.status)
-        if self.code == http.OK:
-            if b'content-length' in headers:
-                self.total_length = int(headers[b'content-length'][0])
-            else:
-                self.total_length = 0
+        self.decoder = None
+        client.Agent.__init__(self, reactor)
+
+    def request_callback(self, response):
+        finished = Deferred()
+
+        if response.code in (
```

Comment:
I was thinking we should be able to use twisted RedirectAgent to deal with redirects and also ContentDecoderAgent for decoding gzip thus simplify this code. 

To deal with deflate might need to subclass GzipDecoder to add 32 instead of 16.

What are your thoughts?

Reply:
hmm, this will simplify things...
But some things will have to the same, like how to handle a redirect when not using `RedirectAgent`. The current code of `TrackerIcon` is using `HTTPDownloader` but handles the redirect of the original page by itself and thous requires a `PageRedirect` if the page marked as redirect.

Reply:
There is the `handle_redirects` arg (I assume Trackicons uses this) for `download_file()` that could be used to determine whether to wrap the agent with `RedirectAgent`?

Reply:
Yes, and in the last commit I use it for this purpose. Just explaining why some of the code must stay.